### DISCUSSION TITLE
Retrieve Named/Provided Dependency in DI Graph

### DISCUSSION THREAD

**Discussion by kevinrobayna** [uber-go/fx#919] [Q&A] [answered]
Hi,

I'm currently writing an app and I wanted to have some integration tests where I could call the dependencies built with fx.

In a Java world with [Guice](http://github.com/google/guice) I would do something like this.

```java
injector.getInstance (Class<T> type);
```

Is there a way to achieve the same with Fx? 

I saw this discussion https://github.com/uber-go/fx/discussions/903 and I saw this piece of code

```Go
func TestServiceB(t *testing.T) {
  app := fxtest.New(t,
    opts(),
    fx.Decorate(newFakeServiceA), // returns ServiceA
  )

  // Starts the app right away, and defers a stop when the test ends.
  defer app.RequireStart().RequireStop()

  // TODO: make a request to the app
}
```
But this would be only if I put an HTTP server to which I do a request. How can I just call the service methods?

This is what I wish I could do.

```Go
func TestServiceX(t *testing.T) {
  app := fxtest.New(t, opts())
  defer app.RequireStart().RequireStop()

  x, _ := app.GetProvided/Named(X)  
  result := x.Foo()
  
  //TODO: Assert result contains the expected result
}
```

**Comment by abhinav** [comment, marked as answer]
Hello! You can use the [fx.Populate](https://pkg.go.dev/go.uber.org/fx#Populate) API to request objects from the container.

```go
var x MyFooer
app := fxtest.New(t,
  opts(),
  fx.Populate(&x),
)
defer app.RequireStart().RequireStop()

result := x.Foo()
```

Note that the argument must be a pointer to the type of value you're looking up.
So if the value inside the container is a `*Foo`, the pointer must be a `**Foo`.

```go
var foo *Foo
fxtest.New(t,
  fx.Provide(NewFoo()), // returns *Foo
  fx.Populate(&foo), // **Foo
)
```

You can provide any number of pointers to fill to `fx.Populate`.

Pointers will be filled when `fx.New`/`fxtest.New` returns.
However, we recommend calling `app.Start`/`app.RequireStart` before using them,
because the object might have a dependency that needs to be started before it's usable.

---

**Design consideration: Why is fx.Populate an argument to fx.New/fxtest.New?**

Fx intentionally does not allow looking up objects in the App after creation.
We want to avoid an anti-pattern: passing the App object to business logic, and looking up values inside it on the request path.

**Reply by kevinrobayna** [reply]
Oh that makes sense!

tbh i agree with you making it accessible on “prod” code isn’t great and leads to anti patterns as you said.

thank you very much!
